### PR TITLE
[DoctrineBridge] Kill DBAL 2 support

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | 7.0
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

DBAL 2 is dead and buried. Let's remove code that we needed to maintain DBAL 2.

TODO:

- [x] Deprecation layer for removed classes
- [x] Check changes for possible backports to 5.4/6.4
- [x] #50576
- [x] #50579
- [x] #50575

### CODE REVIEW

File: src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php [hunk 1]
```diff
@@ -32,29 +31,14 @@ class DoctrineDataCollector extends DataCollector
...
 
-    /**
-     * @var array<string, DebugStack>
-     */
-    private array $loggers = [];
-
     public function __construct(
         private ManagerRegistry $registry,
-        private ?DebugDataHolder $debugDataHolder = null,
+        private DebugDataHolder $debugDataHolder,
```
Comment (stof) [inline src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php:36]: passing `null` needs to be deprecated before removing it

Comment (derrabus) [inline src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php:36]: TODO: Deprecation layer

File: src/Symfony/Bridge/Doctrine/Logger/DbalLogger.php [hunk 2]
```diff
@@ -1,87 +0,0 @@
...
-namespace Symfony\Bridge\Doctrine\Logger;
-
-use Doctrine\DBAL\Logging\SQLLogger;
-use Psr\Log\LoggerInterface;
-use Symfony\Component\Stopwatch\Stopwatch;
-
-/**
- * @author Fabien Potencier <fabien@symfony.com>
- */
-class DbalLogger implements SQLLogger
```
Comment (derrabus) [inline src/Symfony/Bridge/Doctrine/Logger/DbalLogger.php]: TODO: Deprecation layer

File: src/Symfony/Bridge/Doctrine/Tests/DoctrineTestHelper.php [hunk 3]
```diff
@@ -46,26 +46,20 @@ public static function createTestEntityManager(Configuration $config = null): En
 
         $config ??= self::createTestConfiguration();
 
-        if (!(new \ReflectionMethod(EntityManager::class, '__construct'))->isPublic()) {
-            return EntityManager::create($params, $config);
-        }
```
Comment (derrabus) [inline src/Symfony/Bridge/Doctrine/Tests/DoctrineTestHelper.php]: TODO: Backport to 6.4

File: src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php [hunk 4]
```diff
@@ -35,20 +35,14 @@
...
 {
-    private function createExtractor()
+    private function createExtractor(): DoctrineExtractor
     {
         $config = ORMSetup::createConfiguration(true);
         $config->setMetadataDriverImpl(new AttributeDriver([__DIR__.'/../Tests/Fixtures' => 'Symfony\Bridge\Doctrine\Tests\Fixtures'], true));
-        if (class_exists(DefaultSchemaManagerFactory::class)) {
-            $config->setSchemaManagerFactory(new DefaultSchemaManagerFactory());
-        }
+        $config->setSchemaManagerFactory(new DefaultSchemaManagerFactory());
```
Comment (derrabus) [inline src/Symfony/Bridge/Doctrine/Tests/PropertyInfo/DoctrineExtractorTest.php:42]: TODO: Backport to 6.4

File: src/Symfony/Bridge/Doctrine/Tests/Types/UlidTypeTest.php [hunk 5]
```diff
@@ -146,6 +144,7 @@ public static function provideSqlDeclarations(): array
             [new PostgreSQLPlatform(), 'UUID'],
             [new SqlitePlatform(), 'BLOB'],
             [new MySQLPlatform(), 'BINARY(16)'],
+            [new MariaDBPlatform(), 'BINARY(16)'],
```
Comment (derrabus) [inline src/Symfony/Bridge/Doctrine/Tests/Types/UlidTypeTest.php]: TODO: Backport to ~~6.2~~ 5.4.

#50574

File: src/Symfony/Component/Cache/Adapter/DoctrineDbalAdapter.php [hunk 6]
```diff
@@ -71,26 +71,20 @@ public function __construct(Connection|string $connOrDsn, string $namespace = ''
...
+                'db2' => 'ibm_db2',
+                'mssql' => 'pdo_sqlsrv',
+                'mysql' => 'pdo_mysql',
+                'mysql2' => 'pdo_mysql',
+                'postgres' => 'pdo_pgsql',
+                'postgresql' => 'pdo_pgsql',
+                'pgsql' => 'pdo_pgsql',
+                'sqlite' => 'pdo_sqlite',
+                'sqlite3' => 'pdo_sqlite',
+            ]))->parse($connOrDsn);
```
Comment (derrabus) [inline src/Symfony/Component/Cache/Adapter/DoctrineDbalAdapter.php:84]: TODO: Backport to 6.4

File: src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php (hunk 1 above)
Comment (derrabus) [reply src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php:36]: Yes.

File: src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php [hunk 7]
```diff
@@ -32,29 +31,14 @@ class DoctrineDataCollector extends DataCollector
...
         $this->connections = $registry->getConnectionNames();
         $this->managers = $registry->getManagerNames();
     }
 
-    /**
-     * Adds the stack logger for a connection.
-     *
-     * @return void
-     */
-    public function addLogger(string $name, DebugStack $logger)
```
Comment (chalasr) [inline src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php]: TODO: Deprecation layer :)

**Comment by nicolas-grekas** [comment]
Can't we target 6.4? That'd be nicer to the community.

File: src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php (hunk 7 above)
Comment (derrabus) [reply src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php]: yes!

**Comment by derrabus** [comment]
> Can't we target 6.4?

You mean drop DBAL 2 in 6.4 already? We can, but I didn't want to cut off projects that still run on DBAL 2 (or at least support it) on a minor release. I cannot really tell what the impact of that change would be.

> That'd be nicer to the community.

One could argue that giving the community an LTS for migrating away from a still very popular package would be nicer though.

**Comment by stof** [comment]
Well, older versions of DoctrineBundle don't work well with Symfony 6.3 anyway (for instance, the `MapEntity` attribute would break) and DoctrineBundle 2.9+ already requires DBAL 3.6+.
So I would say that if you are still using DBAL 2, you are almost required to stick to Symfony 5.4 currently.

**Comment by nicolas-grekas** [comment]
> You mean drop DBAL 2 in 6.4 already?

Yes!
Major versions should be as boring as possibly doable, and that includes not doing version bumps if possible. Minor versions are better candidates for those bumps.

File: src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php (hunk 1 above)
Comment (derrabus) [reply src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php:36]: Done in #50579

Comment (derrabus) [reply src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php:36]: #50579

File: src/Symfony/Bridge/Doctrine/Logger/DbalLogger.php (hunk 2 above)
Comment (derrabus) [reply src/Symfony/Bridge/Doctrine/Logger/DbalLogger.php]: #50579

File: src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php (hunk 7 above)
Comment (derrabus) [reply src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php]: #50579

File: src/Symfony/Bridge/Doctrine/Tests/DoctrineTestHelper.php (hunk 3 above)
Comment (derrabus) [reply src/Symfony/Bridge/Doctrine/Tests/DoctrineTestHelper.php]: #50576

**Comment by derrabus** [comment]
> Minor versions are better candidates for those bumps.

In our case, it leaves DBAL 2 projects pinned to a non-LTS Symfony branch. But well, [they have been warned](https://www.doctrine-project.org/2022/01/22/sunsetting-dbal-2.html), right?

I can work out a PR for 6.4 once #50576 and #50579 are done. We can continue the discussion there.

**Comment by nicolas-grekas** [comment]
I'm torn now. I'd fine either way, which means fine for 7.0 :)

File: UPGRADE-7.0.md [hunk 8]
```diff
@@ -5,6 +5,13 @@ Symfony 6.4 and Symfony 7.0 will be released simultaneously at the end of Novemb
 release process, both versions will have the same features, but Symfony 7.0 won't include any deprecated features.
 To upgrade, make sure to resolve all deprecation notices.
 
+DoctrineBridge
+--------------
+
+* Remove `DoctrineDbalCacheAdapterSchemaSubscriber`, use `DoctrineDbalCacheAdapterSchemaListener` instead
+* Remove `MessengerTransportDoctrineSchemaSubscriber`, use `MessengerTransportDoctrineSchemaListener` instead
+* Remove `RememberMeTokenProviderDoctrineSchemaSubscriber`, use `RememberMeTokenProviderDoctrineSchemaListener` instead
```
Comment (nicolas-grekas) [inline UPGRADE-7.0.md]: missing space at start of line (+rebase needed)

File: UPGRADE-7.0.md [hunk 9]
```diff
@@ -5,6 +5,13 @@ Symfony 6.4 and Symfony 7.0 will be released simultaneously at the end of Novemb
 release process, both versions will have the same features, but Symfony 7.0 won't include any deprecated features.
 To upgrade, make sure to resolve all deprecation notices.
 
+DoctrineBridge
+--------------
+
+ * Remove `DoctrineDbalCacheAdapterSchemaSubscriber`, use `DoctrineDbalCacheAdapterSchemaListener` instead
+ * Remove `MessengerTransportDoctrineSchemaSubscriber`, use `MessengerTransportDoctrineSchemaListener` instead
+ * Remove `RememberMeTokenProviderDoctrineSchemaSubscriber`, use `RememberMeTokenProviderDoctrineSchemaListener` instead
```
Comment (nicolas-grekas) [inline UPGRADE-7.0.md]: we should say something about DbalLogger + duplicate those lines in the changelog of the bridge, isn't it?

**Review by nicolas-grekas — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @derrabus.

**Pull request merged**